### DISCUSSION TITLE
Coc does not always report about errors? Should I use ALE with Coc?

### DISCUSSION THREAD

**Discussion by vitaly-zdanevich** [neoclide/coc.nvim#4571] [Q&A] [answered]
I write mostly Go, sometimes TypeScript.

I tried to remove ALE, but found that some errors are not reported by Coc.

Do you recommend to use ALE with Coc?

**Comment by fannheyward** [comment]
It's up to you. Some language server doesn't report as many diagnostics as ale linters, for example, there're `golang-lint`, `static check` for Go, they can report more diagnostics than `gopls`.

You can use coc.nvim with ALE, additional, you can display coc.nvim's diagnostics with ALE `:h coc-config-diagnostic-displayByAle`.

**Reply by vitaly-zdanevich** [reply]
golang-lint, static check - is it possible to enable them in CoC?

**Reply by sisrfeng** [reply, marked as answer]
From new bing:


Hello, this is Bing. I can help you find some information about linting ALE vs coc. 😊

Linting is the process of checking the source code for errors, bugs, stylistic issues, and other problems. ALE and coc are two popular plugins for Vim that provide linting and other features, such as code completion, formatting, and refactoring.

ALE stands for **Asynchronous Lint Engine**¹, which means it runs external linter commands in the background and displays the results in the editor. ALE supports more than 200 languages and tools¹, and can be configured to use different linters and fixers for each file type. ALE also supports some features of the Language Server Protocol (LSP), such as go-to-definition, hover information, and code actions¹.

coc stands for **Conquer of Completion**², which is a plugin that aims to provide a full-featured development environment in Vim. coc relies on node.js and npm to install and run extensions that provide language-specific features². coc supports many LSP features, such as code completion, diagnostics, formatting, renaming, signature help, and more². coc also has a built-in fuzzy finder and snippet engine².

The choice between ALE and coc depends on your personal preference and needs. Some possible factors to consider are:

- **Performance**: ALE may be faster and smoother than coc, as it does not require node.js or npm, and it runs linters asynchronously without blocking the editor³. However, coc may provide more accurate and consistent results, as it uses dedicated language servers that run continuously and communicate with the editor via JSON-RPC⁴.
- **Features**: coc may offer more features than ALE, as it supports many extensions that provide rich functionality for various languages and frameworks². However, ALE may support more linters and fixers than coc, as it can run any external command that produces output in a standard format¹.
- **Configuration**: ALE may be easier to configure than coc, as it uses Vim options and variables to set up linters and fixers¹. However, coc may be more flexible than ALE, as it allows you to customize the behavior of each extension using JSON files².

You can also use both plugins together, if you want to combine the strengths of each one. For example, you can use ALE for linting and fixing, and use coc for code completion and other LSP features⁵. You can also blacklist some languages or tools in ALE or coc, to avoid conflicts or redundancy⁵.

I hope this helps you understand the differences between ALE and coc. If you have any more questions, feel free to ask me. 😊

¹: [ALE - Asynchronous Lint Engine](^[1](https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/)^)
²: [coc.nvim - Intellisense engine for Vim8 & Neovim](^2^)
³: [CoC vs ALE or do you need both or is one enough? : r/vim - Reddit](^4^)
⁴: [What is the advantage of ALE over CoC? : r/vim - Reddit](^[1](https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/)^)
⁵: [Using both CoC & ALE together? : r/vim - Reddit](https://www.reddit.com/r/vim/comments/9z8x0t/using_both_coc_ale_together/)

Source: Conversation with Bing, 8/2[1](https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/)/[2](https://www.reddit.com/r/vim/comments/ldiwlo/coc_vs_ale_or_do_you_need_both_or_is_one_enough/)0[2](https://www.reddit.com/r/vim/comments/ldiwlo/coc_vs_ale_or_do_you_need_both_or_is_one_enough/)3
(1) What is the advantage of ALE over CoC? : r/vim - Reddit. https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/.
([2](https://www.reddit.com/r/vim/comments/ldiwlo/coc_vs_ale_or_do_you_need_both_or_is_one_enough/)) CoC vs ALE or do you need both or is one enough? : r/vim - Reddit. https://www.reddit.com/r/vim/comments/ldiwlo/coc_vs_ale_or_do_you_need_both_or_is_one_enough/.
([3](https://stackoverflow.com/questions/71357070/using-ale-with-clangd-for-both-linting-and-autocompletion)) Using ALE with clangd for both linting and autocompletion. https://stackoverflow.com/questions/71357070/using-ale-with-clangd-for-both-linting-and-autocompletion.
([4](https://www.reddit.com/r/vim/comments/dd010c/linting_cocvim_or_ale/)) Linting: coc.vim or ALE? : r/vim - Reddit. https://www.reddit.com/r/vim/comments/dd0[1](https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/)0c/linting_cocvim_or_ale/.
(5) undefined. https://github.com/Blacksuan[1](https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/)9/init.nvim.
(6) What is the advantage of ALE over CoC? : r/vim - Reddit. https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/.
(7) CoC vs ALE or do you need both or is one enough? : r/vim - Reddit. https://www.reddit.com/r/vim/comments/ldiwlo/coc_vs_ale_or_do_you_need_both_or_is_one_enough/.
(8) Linting: coc.vim or ALE? : r/vim - Reddit. https://www.reddit.com/r/vim/comments/dd0[1](https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/)0c/linting_cocvim_or_ale/.
(9) [What is the advantage of ALE over CoC? : r/vim - Reddit](https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/). https://www.reddit.com/r/vim/comments/fziffu/what_is_the_advantage_of_ale_over_coc/.
(10) undefined. https://github.com/Roco-scientist/vim_neovim_settings.
